### ISSUE TITLE
Fall back to GLES2 automatically on known problematic GPUs with poor GLES3 support

### ISSUE THREAD

**Issue by Calinou** [godotengine/godot-proposals#1609]
**Describe the project you are working on:**

The Godot editor :slightly_smiling_face: 

**Describe the problem or limitation you are having in your project:**

Many desktop and mobile GPUs *claim* to support OpenGL 3.3/OpenGL ES 3.0, but do so poorly in practice. They can also have limitations that are incompatible with the current GLES3 renderer design.

Here are some examples of such GPUs I've listed over the years, based on issue reports:

### Desktop

- AMD Radeon HD 3000/4000/5000 series
- NVIDIA GeForce 8000/9000 series
- 3rd-generation Intel HD Graphics (Ivy Bridge)

### Mobile

- Many Adreno GPUs

*Note: I don't have issue links at hand, but it should be possible to find many of them by searching for [`is:issue is:open label:topic:rendering gles3`](https://github.com/godotengine/godot/issues?q=is%3Aissue+is%3Aopen+gles3+label%3Atopic%3Arendering) in the Godot issue tracker.*

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**

If the project uses GLES3 and **Fallback to Gles2** is enabled in the project settings, use GLES2 if the GPU is on the exclusion list.

This should be implemented in the `3.2` branch first, as the current `master` branch doesn't have GLES3 or GLES2 renderers yet.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**

- Store a list of substrings to detect in `VisualServer::get_video_adapter_name()` somewhere in the engine's source code. This list generally don't need to change over time, since new GPU models are unlikely to have poor GLES3 support.
- If there is a match in any of the substrings, fall back to GLES2 automatically.
- Change the`--video-driver` command line argument to work as follows:
  - If not specified, use the project-defined video driver. If GLES3 is used and "Fallback To Gles2" is enabled in the Project Settings, allow falling back to GLES2 on unsupported GPUs (or GPUs on the aforementioned exclusion list).
  - If `--video-driver GLES3` is used, always attempt to use GLES3 even if the GPU is on the exclusion list. This can be used to bypass the automatic fallback, just in case.

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**

No, as the renderer choice happens before any of the project's code is run.

**Is there a reason why this should be core and not an add-on in the asset library?:**

This is core engine functionality, and is required to improve the user experience out of the box.

**Comment by salvprest** [comment]
I think that this is the best way to handle this kinds of broken gpu drivers, on mobile this is an unresolved problem

**Comment by Calinou** [comment]
I started working on this on the `3.x` branch: https://github.com/Calinou/godot/tree/fallback-gles2-specific-gpus

The problem is that falling back to GLES2 this way will make the project manager not draw anything other than the clear color, and OpenGL errors will be spammed to the console:

```
❯ bin/godot.x11.tools.64.llvm --verbose
Godot Engine v3.3.rc.custom_build.00d087e47 - https://godotengine.org
XInput: Refreshing devices.
XInput: No touch devices found.
Detecting GPUs, set DRI_PRIME in the environment to override GPU detection logic.
Only one GPU found, using default.
GeForce GTX 1080/PCIe/SSE2
Using GLES2 video driver
OpenGL ES 2.0 Renderer: GeForce GTX 1080/PCIe/SSE2
ERROR: _gl_debug_print: GL ERROR: Source: OpenGL        Type: Error     ID: 1280        Severity: High  Message: GL_INVALID_ENUM error generated. Operation is not valid from the core profile.
   At: drivers/gles2/rasterizer_gles2.cpp:133.
ERROR: _gl_debug_print: GL ERROR: Source: OpenGL        Type: Error     ID: 1281        Severity: High  Message: GL_INVALID_VALUE error generated. <internalFormat> not valid.
   At: drivers/gles2/rasterizer_gles2.cpp:133.
ERROR: _gl_debug_print: GL ERROR: Source: OpenGL        Type: Error     ID: 1280        Severity: High  Message: GL_INVALID_ENUM error generated. Cannot enable <cap> in the current profile.
   At: drivers/gles2/rasterizer_gles2.cpp:133.
OpenGL ES Batching: ON
        OPTIONS
        max_join_item_commands 16
        colored_vertex_format_threshold 0.25
        batch_buffer_size 16384
        light_scissor_area_threshold 1
        item_reordering_lookahead 4
        light_max_join_items 32
        single_rect_fallback False
        debug_flash False
        diagnose_frame False
PulseAudio: detected 2 channels
PulseAudio: audio buffer frames: 512 calculated latency: 11ms
JoypadLinux: udev enabled and loaded successfully.
 
ERROR: _gl_debug_print: GL ERROR: Source: OpenGL        Type: Error     ID: 1282        Severity: High  Message: GL_INVALID_OPERATION error generated. Invalid VAO/VBO/pointer usage.
   At: drivers/gles2/rasterizer_gles2.cpp:133.
ERROR: _gl_debug_print: GL ERROR: Source: OpenGL        Type: Error     ID: 1282        Severity: High  Message: GL_INVALID_OPERATION error generated. Array object is not active.
   At: drivers/gles2/rasterizer_gles2.cpp:133.
ERROR: _gl_debug_print: GL ERROR: Source: OpenGL        Type: Error     ID: 1281        Severity: High  Message: GL_INVALID_VALUE error generated. <internalFormat> not valid.
   At: drivers/gles2/rasterizer_gles2.cpp:133.
ERROR: _gl_debug_print: GL ERROR: Source: OpenGL        Type: Error     ID: 1281        Severity: High  Message: GL_INVALID_VALUE error generated. <internalFormat> not valid.
   At: drivers/gles2/rasterizer_gles2.cpp:133.
ERROR: _gl_debug_print: GL ERROR: Source: OpenGL        Type: Error     ID: 1281        Severity: High  Message: GL_INVALID_VALUE error generated. <internalFormat> not valid.
   At: drivers/gles2/rasterizer_gles2.cpp:133.
ERROR: _gl_debug_print: GL ERROR: Source: OpenGL        Type: Error     ID: 1281        Severity: High  Message: GL_INVALID_VALUE error generated. <internalFormat> not valid.
   At: drivers/gles2/rasterizer_gles2.cpp:133.
```

There is also some commented out code that I added to try to restart Godot with the GLES2 renderer. This way is hackier, but it could be investigated if we can't get this to work.

**Comment by jamie-pate** [comment]
I'd like to add "Intel(R) HD Graphics*" to this list on windows I've seen some complete lockup situations with these chips.
also "Intel Iris Plus Graphics" on macs :( https://github.com/godotengine/godot/issues/52722

**Comment by Calinou** [comment]
> I'd like to add "Intel(R) HD Graphics*" to this list on windows I've seen some complete lockup situations with these chips. also "Intel Iris Plus Graphics" on macs :( [godotengine/godot#52722](https://github.com/godotengine/godot/issues/52722)

Making *all* Intel IGPs on Windows fall back to GLES2 would be kind of extreme. Recent IGP models can run GLES3 just fine.

**Comment by jamie-pate** [comment]
HD vs UHD.. they are apparently different enough that the *hd* version crashes even though Wikipedia says that hd 620 and uhd 620 are the same generation,..
I think this may need be to be a blocklist setting?

If there is a sane default then each developer could use that, and pile on any other gpu strings that are flaking out in their specific use case.

The real problem is that developers may not have the resources available to test all these variations. Different parts are flakey for each driver too.

And e.g. apple feedback straight up said to me "GL is not greatly supported, we recommend that you move to Metal"

**Comment by Calinou** [comment]
> Apple feedback straight up said to me "GL is not greatly supported, we recommend that you move to Metal". This doesn't give me much confidence that they care enough about open gl to clean up this situation.

Using [MetalANGLE](https://github.com/godotengine/godot/pull/50253) is likely the solution here.

**Comment by jamie-pate** [comment]
Yes, i noticed that for webgl Chrome just uses ANGLE as well... Seems they were fed up with the situation with Windows OpenGL driver quality as well..

**Comment by jamie-pate** [comment]
FYI Getting reports today that 10th gen `Intel(R) UHD Graphics` are possibly also broken in GLES3

**Comment by Calinou** [comment]
> FYI Getting reports today that 10th gen `Intel(R) UHD Graphics` are possibly also broken in GLES3

We'd need a minimal reproduction project with someone available for testing to be able to confirm this. Otherwise, there's nothing we can do about it (GLES2 may as well be broken on such hardware, if GLES3 is).

**Comment by jamie-pate** [comment]
Yes, that's the problem though, getting the hardware. (And getting users to
cooperate and report what their hardware is doing)

Hoping the vulkan situation is better for Intel.... And maybe the
'compatibility focused' gles3 impl will also be fine in godot4?

This is one low quality signal, but it points to the requirement for adding
this feature as a configuration setting so that each product can set their
own compatibility list. Different games use different features and with the
feature set I'm using I have a very low confidence in the gles3 driver
quality of amost all intel GPUs.

**Comment by jamie-pate** [comment]
Intel says:
`We recommend contacting the developer of the app in order to consider listing the app as compatible only with Iris Xe Graphics and the newer versions.`

See https://github.com/godotengine/godot/issues/56496#issuecomment-1021449560

**Comment by jamie-pate** [comment]
> GLES2 may as well be broken on such hardware, if GLES3 is

GLES2 is 100% fine afaict on all intel igpus on all platforms.
